<!DOCTYPE html>
<html lang="en">
<head>
        <meta charset="utf-8">
        <title>Grad Hustle - 454</title>
        <link rel="stylesheet" href="/theme/css/main.css">
                
        <!--[if IE]>
            <script src="http://html5shiv.googlecode.com/svn/trunk/html5.js"></script>
        <![endif]-->
</head>

<body id="index" class="home">
        <header id="banner" class="body">
                <h1><a href="/">Grad Hustle </a></h1>
                <nav><ul>
                                                                                    <li ><a href="/category/linuxunix.html">Linux/Unix</a></li>
                                                </ul></nav>
        </header><!-- /#banner -->
                
            

                            <aside id="featured" class="body">
                <article>
                    <h1 class="entry-title"><a href="/destroying-angel-genome-assembly-with-pacbio-sequences.html">Destroying Angel Genome Assembly with PacBio Sequences</a></h1> 
                    <footer class="post-info">
        <abbr class="published" title="2013-07-24T01:53:00">
                Wed 24 July 2013
        </abbr>

                <address class="vcard author">
                By <a class="url fn" href="/author/likit-preeyanon.html">Likit Preeyanon</a>
        </address>
        <p>In <a href="/category/linuxunix.html">Linux/Unix</a>. </p>
<p>tags: <a href="/tag/ngs.html">NGS</a><a href="/tag/genome-assembly.html">Genome assembly</a><a href="/tag/pacbio.html">PacBio</a><a href="/tag/454.html">454</a><a href="/tag/smrt.html">SMRT</a></p>
</footer><!-- /.post-info --><h1>Background</h1>
<p>I have been collaborating with <a href="http://waltonlab.prl.msu.edu/labmembers/jonathan-walton">Jonathan Walton</a>,
a professor in Plant biology department at MSU on genome assembly of a toxic mushroom called
destroying angel (<em>A. bisporigera</em>).
Right now, we have 9 datasets from PacBio in bas.h5 format (a movie file).
The file size is about 1.5-3.0 Gb.
Moreover, there are also files in fasta, fastq and ccs.fastq. However, I learned later that those
files were useless because they were not quality filtered and still contained adapters.</p>
<h1>Quality Filtering and Trimming with SMRT PreAssembly Pipeline</h1>
<p>SMRT software is available on github <a href="https://github.com/PacificBiosciences/SMRT-Analysis/wiki/SMRT-Pipe-Reference-Guide-v2.0">here</a>
and also pre-installed on EC2 machine (ami-fbbbcf92), so I launched EC2 instance and learned to run the software.</p>
<p>The software has several pipelines, each consists of several modules.
You can also build your own pipeline by putting the modules together.
I ran <strong>AlloraSFilter</strong> module on the command line with smrtpipe.py with quality score 0.75 to all bas.h5 files.
Here are the results.</p>
<p>As you can see, the polymerase read length increases from 2639 bp  to 6019 bp and the mean quality score also
increases from 0.30 to 0.83. The plots below show a histogram of read lengths and quality scores of polymerase reads.</p>
<p><img alt="lengths_histogram" src="/images/postFilterRLHistogram.png" /></p>
<p><img alt="scores_histogram" src="/images/postFilterRSHistogram.png" /></p>
<p>And here is a histogram of subread lengths (after adaptors are removed).</p>
<p><img alt="subreads_histogram" src="/images/filteredSubreadReport.png" /></p>
<p>The module also generated fastq and fasta file for both subreads and consensus reads (CCS) from bas.h5 files.</p>
<h1>Error Correction and Assembly</h1>
<p>After that, I ran pacToCA to error correct subreads with CCS sequences and run Celera assembler to assembled corrected reads.
Celera produced five different output files, singletons, unitigs, contigs, degenerate contigs and scaffolds.
Basically, it built unique contigs from reads, and reads not used to build unique contigs were stored in singletons file.
Then unique contigs were aligned to build contigs. Unique contigs that were not aligned to any other contigs were stored in degenerate contigs file.
Therefore, the contigs file contains high confident contigs (multiple alignments) and degenerate file contains low confident contigs.
Scaffolds contain scaffolded contigs; however, in our case no contigs were scaffolded, so it contains exactly the same contigs in contigs file.</p>
<p>Unfortunately, the total length of high confident contigs is only 6.4Mbp whereas that of low confident contigs is about 8.9Mbp.
The max contig size is 13,072 bp and the max size of degenerate contig is 78,021. N50 sizes are 2,147 bp and 997 bp respectively.
I know the estimated size of the genome is ~50 Mbp, so this really low. However, I learned that the overall assembly size is
comparable to that of the assembly done by other collaborators.
So, the issue could be from DNA samples or sequencing, not from the assembly.</p>
<p>Also, I tried to combined 454 assembly that was done by someone with PacBio assembly from Celera assembly.
I combined singletons, unique contigs, degenerate contigs and 454 contigs and assembled them with CAP3 genome assembly.</p>
<p>The results show that the total contig size is a lot bigger (17.5 Mbp) and total size of singleton contigs is about 21 Mbp.
Max contig size also increases (24 Kbp) as well as N50 size (1.4 Kbp).</p>
<h1>Conclusion</h1>
<p>It seems like mushroom genome with PacBio runs into the same problem with other genome assembly projects.
A lot of reads and contigs (even with length = 78 Kbp) cannot be assembled into long contigs.
It also appears that 454 assembly also have unique contigs that cannot be merged to PacBio assembly.
These problems may be solved by paired-end with different insert sizes and Illumina sequencing.</p><p>There are <a href="/destroying-angel-genome-assembly-with-pacbio-sequences.html#disqus_thread">comments</a>.</p>                </article>
                                    <p class="paginator">
        Page 1 / 1
    </p>
                            </aside><!-- /#featured -->
                                                </ol><!-- /#posts-list -->
                        </section><!-- /#content -->
                    <section id="extras" class="body">
                        <div class="blogroll">
                        <h2>blogroll</h2>
                        <ul>
                                                    <li><a href="http://getpelican.com/">Pelican</a></li>
                                                    <li><a href="http://python.org/">Python.org</a></li>
                                                    <li><a href="http://jinja.pocoo.org/">Jinja2</a></li>
                                                    <li><a href="http://ged.msu.edu">GED Lab</a></li>
                                                </ul>
                </div><!-- /.blogroll -->
                                <div class="social">
                        <h2>social</h2>
                        <ul>
                            <li><a href="/None" type="application/atom+xml" rel="alternate">atom feed</a></li>
                            
                                                    <li><a href="#">You can add links in your config file</a></li>
                                                    <li><a href="#">Another social link</a></li>
                                                </ul>
                </div><!-- /.social -->
                </section><!-- /#extras -->

        <footer id="contentinfo" class="body">
                <address id="about" class="vcard body">
                Proudly powered by <a href="http://getpelican.com/">Pelican</a>, which takes great advantage of <a href="http://python.org">Python</a>.
                </address><!-- /#about -->

                <p>The theme is by <a href="http://coding.smashingmagazine.com/2009/08/04/designing-a-html-5-layout-from-scratch/">Smashing Magazine</a>, thanks!</p>
        </footer><!-- /#contentinfo -->

<script type="text/javascript">
    var disqus_shortname = 'likitsblog';
    (function () {
        var s = document.createElement('script'); s.async = true;
        s.type = 'text/javascript';
        s.src = 'http://' + disqus_shortname + '.disqus.com/count.js';
        (document.getElementsByTagName('HEAD')[0] || document.getElementsByTagName('BODY')[0]).appendChild(s);
    }());
</script>
</body>
</html>