<!DOCTYPE html>
<html lang="en">
<head>
        <meta charset="utf-8">
        <title>Grad Hustle - Illumina</title>
        <link rel="stylesheet" href="/theme/css/main.css">
                
        <!--[if IE]>
            <script src="http://html5shiv.googlecode.com/svn/trunk/html5.js"></script>
        <![endif]-->
</head>

<body id="index" class="home">
        <header id="banner" class="body">
                <h1><a href="/">Grad Hustle </a></h1>
                <nav><ul>
                                                                                    <li ><a href="/category/linuxunix.html">Linux/Unix</a></li>
                                    <li ><a href="/category/next-gen-sequencing.html">Next-Gen Sequencing</a></li>
                                                </ul></nav>
        </header><!-- /#banner -->
                
            

                            <aside id="featured" class="body">
                <article>
                    <h1 class="entry-title"><a href="/running-bowtiebowtie2-rsem-and-tophat-on-dutp-strand-specific-reads.html">Running Bowtie/Bowtie2, RSEM and Tophat on dUTP Strand-Specific Reads</a></h1> 
                    <footer class="post-info">
        <abbr class="published" title="2013-10-22T12:03:00">
                Tue 22 October 2013
        </abbr>

                <address class="vcard author">
                By <a class="url fn" href="/author/likit-preeyanon.html">Likit Preeyanon</a>
        </address>
        <p>In <a href="/category/next-gen-sequencing.html">Next-Gen Sequencing</a>. </p>
<p>tags: <a href="/tag/ngs.html">NGS</a><a href="/tag/illumina.html">Illumina</a><a href="/tag/rnaseq.html">RNASeq</a><a href="/tag/bowtie.html">Bowtie</a><a href="/tag/rsem.html">RSEM</a><a href="/tag/tophat.html">Tophat</a></p>
</footer><!-- /.post-info --><h1>Background</h1>
<p>I had been working on strand-specific paired-end reads from HiSeq lately and
I had trouble mapping reads back to assembled transcripts using Bowtie
as well as using RSEM to estimate transcript abundance.
After some google searches, I found some information about the dUTP method used to generate stranded reads on
<a href="http://support.illumina.com/sequencing/sequencing_kits/truseq_stranded_mrna_lt_sample_prep_kit/questions.ilmn">Illumina website</a>
as well as a <a href="http://onetipperday.blogspot.com/2012/07/how-to-tell-which-library-type-to-use.html">blog</a>
that explains an orientation of reads from different protocols and which flag (library type) should be used to run Tophat.
However, I could not find any information on how to run Bowtie/Bowtie2 as well as RSEM, which uses Bowtie as a default aligner.
After a couple days of frustration here is what I learned.</p>
<h1>Bowtie and Tophat flags for strand-specific reads</h1>
<p>Tophat uses <strong>--fr-firststrand</strong> for a library created by the dUTP method.
This is stated clearly in the manual, so it is easy to understand.
In contrast, Bowtie/Bowtie2 uses <strong>--fr</strong>, <strong>--rf</strong>, <strong>--ff</strong> to specify the orientation of paired-end reads.</p>
<p><strong>--fr</strong> means the upstream read (/1) is from a forward strand and the downstream read (/2) is from a reverse strand.</p>
<p><strong>--rf</strong> means the upstream read (/1) is from a reverse strand and the downstream read (/2) is from a forward strand.</p>
<p><strong>--ff</strong> means both reads are from a forward strand.</p>
<p><img alt="paired-end orientation" src="/static/images/pe-orient.png" /></p>
<p>Credit Xianjun Dong.</p>
<p>With <strong>--fr</strong> flag, the upstream reads could be either /1 and /2, which is valid for unstranded reads.
However, as you can see from the figure, the /1 reads from dUTP are from the reverse strand only.
So, we need to tell Bowtie <strong>not to</strong> map the /1 reads to the forward strand.
This behavior can be achieved by specifying <strong>--nofw</strong> flag.</p>
<p>Therefore, to run Bowtie on dUTP reads, the command should look something like this:</p>
<div class="highlight"><pre><span class="n">bowtie</span> <span class="o">-</span><span class="n">S</span> <span class="o">-</span><span class="n">X</span> <span class="mi">1000</span> <span class="o">--</span><span class="n">fr</span> <span class="o">--</span><span class="n">nofw</span> <span class="n">bowtie</span><span class="o">-</span><span class="n">index</span> <span class="o">-</span><span class="mi">1</span> <span class="n">reads_R1</span><span class="p">.</span><span class="n">fastq</span> <span class="o">-</span><span class="mi">2</span> <span class="n">reads_R2</span><span class="p">.</span><span class="n">fastq</span> <span class="o">&gt;</span> <span class="n">output</span><span class="p">.</span><span class="n">sam</span>
</pre></div>


<p>To compare the results, I ran Bowtie/1.0.0 on the same dataset (1M reads) with different flags specified.</p>
<table>
<thead>
<tr>
<th><strong>Flag</strong></th>
<th><strong>Mapped Reads</strong></th>
<th><strong>Unmapped Reads</strong></th>
</tr>
</thead>
<tbody>
<tr>
<td>--fr (default)</td>
<td>82.57%</td>
<td>17.43%</td>
</tr>
<tr>
<td>--rf</td>
<td>0.17%</td>
<td>99.83%</td>
</tr>
<tr>
<td>--ff</td>
<td>0.00%</td>
<td>100.00%</td>
</tr>
<tr>
<td>--fr --nofw</td>
<td>82.18%</td>
<td>17.82%</td>
</tr>
<tr>
<td>--fr --norc</td>
<td>1.99%</td>
<td>98.01%</td>
</tr>
</tbody>
</table>
<p>The number of reads mapped with <strong>--fr</strong> and <strong>--nofw</strong> is a bit lower than that from <strong>--fr</strong> alone.
This is due to the fact that without <strong>--nofw</strong> flag, Bowtie maps /1 reads to both strands.</p>
<p>Note, with <strong>--norc</strong> flag, Bowtie will not attempt to map the /1 reads to the reverse strand.</p>
<h1>Running RSEM on dUTP reads</h1>
<p>RSEM has <strong>--strand-specific</strong> flag that will turn on Bowtie <strong>--norc</strong> flag, which will force Bowtie to try to map
the /1 reads to the forward strand only. This is suitable for reads generated by the ligation method.
Also, the <strong>--strand-specific</strong> flag is equivalent to setting <strong>--forward-prob flag</strong> to <strong>1</strong>.</p>
<p>However, for dUTP reads, the <strong>--forward-prob</strong> flag should be set to <strong>0</strong> instead.
It will specify <strong>--nofw</strong> flag, which forces Bowtie to map the /1 reads to the reverse strand.
Consequently, the <strong>--strand-specific</strong> flag should not be used when <strong>--forward-prob</strong> is set to <strong>0</strong>.
The <strong>--forward-prob</strong> flag not only specifies <strong>--nofw</strong> and <strong>--norc</strong> for Bowtie, but also modifies the orientation
parameters of RSEM model.</p>
<p>For example, the command should look something like this:</p>
<div class="highlight"><pre><span class="n">rsem</span><span class="o">-</span><span class="n">calculate</span><span class="o">-</span><span class="n">expression</span> <span class="o">-</span><span class="n">p</span> <span class="mi">8</span> <span class="o">--</span><span class="n">forward</span><span class="o">-</span><span class="n">prob</span> <span class="mi">0</span> <span class="o">--</span><span class="n">paired</span><span class="o">-</span><span class="n">end</span> <span class="n">sample_r1</span><span class="p">.</span><span class="n">fastq</span> <span class="n">sample_r2</span><span class="p">.</span><span class="n">fastq</span> <span class="n">index</span> <span class="n">sample_output</span>
</pre></div>


<p>New flags for strand-specific paired-end reads will be added to RSEM in the next minor release.
You can read the discussion on this issue <a href="https://groups.google.com/forum/#!msg/rsem-users/SdR0VUOP7l4/oexuV9y1BvgJ">here</a>.</p>
<h1>Trinity Assembly</h1>
<p>To run Trinity with dUTP reads, you need to use <strong>RF</strong> for <strong>--SS_lib_type</strong> flag, which means the /1 reads are from the reverse strand and
the /2 reads are from the forward strand.
Note that, <strong>RF</strong> flag in Trinity is not the same as <strong>--rf</strong> in Bowtie.</p><p>There are <a href="/running-bowtiebowtie2-rsem-and-tophat-on-dutp-strand-specific-reads.html#disqus_thread">comments</a>.</p>                </article>
                                    <p class="paginator">
        Page 1 / 1
    </p>
                            </aside><!-- /#featured -->
                                                </ol><!-- /#posts-list -->
                        </section><!-- /#content -->
                    <section id="extras" class="body">
                        <div class="blogroll">
                        <h2>blogroll</h2>
                        <ul>
                                                    <li><a href="http://getpelican.com/">Pelican</a></li>
                                                    <li><a href="http://python.org/">Python.org</a></li>
                                                    <li><a href="http://jinja.pocoo.org/">Jinja2</a></li>
                                                    <li><a href="http://ged.msu.edu">GED Lab</a></li>
                                                </ul>
                </div><!-- /.blogroll -->
                                <div class="social">
                        <h2>social</h2>
                        <ul>
                            <li><a href="/None" type="application/atom+xml" rel="alternate">atom feed</a></li>
                            
                                                    <li><a href="#">You can add links in your config file</a></li>
                                                    <li><a href="#">Another social link</a></li>
                                                </ul>
                </div><!-- /.social -->
                </section><!-- /#extras -->

        <footer id="contentinfo" class="body">
                <address id="about" class="vcard body">
                Proudly powered by <a href="http://getpelican.com/">Pelican</a>, which takes great advantage of <a href="http://python.org">Python</a>.
                </address><!-- /#about -->

                <p>The theme is by <a href="http://coding.smashingmagazine.com/2009/08/04/designing-a-html-5-layout-from-scratch/">Smashing Magazine</a>, thanks!</p>
        </footer><!-- /#contentinfo -->

<script type="text/javascript">
    var disqus_shortname = 'likitsblog';
    (function () {
        var s = document.createElement('script'); s.async = true;
        s.type = 'text/javascript';
        s.src = 'http://' + disqus_shortname + '.disqus.com/count.js';
        (document.getElementsByTagName('HEAD')[0] || document.getElementsByTagName('BODY')[0]).appendChild(s);
    }());
</script>
</body>
</html>